PR closed - wrong email: restic check - enable --read-data-* with snapshot filter

wplapper opened the pull request:
Enable restic check based on snapshots. The standard snapshot filtering routine  "FindFilteredSnapshots" is used for snapshots selection. However if no snapshots are specified, the subcommand behaves as before.

<!--
Thank you very much for contributing code or documentation to restic! Please
fill out the following questions to make it easier for us to review your
changes.
-->

What does this PR change? What problem does it solve?
-----------------------------------------------------
The PR solves issue 3326. 
<!--
Describe the changes and their purpose here, as detailed as needed.
-->
Add standard snapshot filtering to subcommand.
Add code to cmd/restic/cmd_check.go to detect snapshot filtering 
Add code to internal/checker/checker.go to collect data blobs based on tree visitor. The functions CountPacks() and GetPacks() have been modified, so that snapshot filtering creates a subset of packfiles on which all size calculations will be based.

When snapshot filtering is activated, the complete tree for this snapshot is read and the data blobs - and their packfiles - are extracted.

Was the change previously discussed in an issue or on the forum?
----------------------------------------------------------------
https://github.com/restic/restic/issues/3326

<!--
Link issues and relevant forum posts here.

If this PR resolves an issue on GitHub, use "Closes #1234" so that the issue
is closed automatically when this PR is merged.
-->
Closes #3326

Checklist
---------

<!--
You do not need to check all the boxes below all at once. Feel free to take
your time and add more commits. If you're done and ready for review, please
check the last box. Enable a checkbox by replacing [ ] with [x].

Please always follow these steps:
- Read the [contribution guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches).
- Enable [maintainer edits](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- Run `gofmt` on the code in all commits.
- Format all commit messages in the same style as [the other commits in the repository](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits).
-->

- [x] I have added tests for all code changes.
- [x] I have added documentation for relevant changes (in the manual).
- [x] There's a new file in `changelog/unreleased/` that describes the changes for our users (see [template](https://github.com/restic/restic/blob/master/changelog/TEMPLATE)).
- [x] I'm done! This pull request is ready for review.

wplapper committed:
check: enable --read-data-* for specified snapshots - rebase step 1

Add code to cmd/restic/cmd_check.go to detect snapshots
Resolved conflict for cmd/restic/cmd_check.go - runCheck
integrated newCheckCommand(...)

wplapper committed:
check: run check of packfiles filtered via snapshotfilter - rebase part 2

Added code for selecting multiple snapshots.
Added message how many packfiles and their cumulative size were selected.
In internal/checker/checker.go replaced the datablob / packfile selection from using walker.Walk
to restic.StreamTrees - parallelizing the packfile selection.

resolved conflict in cmd_check: allow check for snapshot filter

MichaelEischer commented on cmd/restic/cmd_check.go:
```diff
@@ -352,6 +351,30 @@ func runCheck(ctx context.Context, opts CheckOptions, gopts GlobalOptions, args
 
+	filterBySnapshot := false
+	if len(args) > 0 || !opts.SnapshotFilter.Empty() {
+		snapshotLister, err := restic.MemorizeList(ctx, repo, restic.SnapshotFile)
```
`(c *Checker) LoadSnapshots(...)` has already collected the full snapshot list. Having to read it a second time, means that the code is in the wrong place.

wplapper replied:
moved new code to internal/checker.go

MichaelEischer commented on cmd/restic/cmd_check.go:
```diff
@@ -352,6 +351,30 @@ func runCheck(ctx context.Context, opts CheckOptions, gopts GlobalOptions, args
+		}
+
+		visitedTrees := restic.NewIDSet()
+		for sn := range FindFilteredSnapshots(ctx, snapshotLister, repo, &opts.SnapshotFilter, args) {
```
Should directly use `(&opts.SnapshotFilter).FindAll(...)` instead. The error handling of `FindFilteredSnapshots` is pretty much non-existent.

MichaelEischer commented on internal/checker/checker.go:
```diff
@@ -431,12 +433,24 @@ func (c *Checker) UnusedBlobs(ctx context.Context) (blobs restic.BlobHandles, er
 // CountPacks returns the number of packs in the repository.
 func (c *Checker) CountPacks() uint64 {
-	return uint64(len(c.packs))
+	if len(c.packSet) == 0 {
```
This is just waiting to cause problems. Instead just let `FindDataPackfiles` replace the `packs` map.

wplapper replied:
done so.

MichaelEischer commented on internal/checker/checker.go:
```diff
@@ -522,3 +536,48 @@ func (c *Checker) ReadPacks(ctx context.Context, packs map[restic.ID]int64, p *p
+// Find data packfiles for repository checking  based on snapshots.
+// Use restic.StreamTrees to gather all data blobs and convert them to their
+// containing packfile
+func (c *Checker) FindDataPackfiles(ctx context.Context, repo *repository.Repository, sn *restic.Snapshot,
```
Change this to receive _all_ relevant snapshots at once. Then there's no need for `c.packSet`.

MichaelEischer commented on internal/checker/checker.go:
```diff
@@ -522,3 +536,48 @@ func (c *Checker) ReadPacks(ctx context.Context, packs map[restic.ID]int64, p *p
+		return nil
+	})
+
+	err := wg.Wait()
+	if err != nil {
+		return err
+	}
+
+	return nil
```
```suggestion
	return wg.Wait()
```

wplapper replied:
done so.

MichaelEischer commented on internal/checker/checker.go:
```diff
@@ -522,3 +536,48 @@ func (c *Checker) ReadPacks(ctx context.Context, packs map[restic.ID]int64, p *p
+				for _, content := range node.Content {
+					result := repo.LookupBlob(restic.DataBlob, content)
+					if len(result) == 0 {
+						return fmt.Errorf("checker.LookupBlob: datablob %s not mapped!", content.Str())
```
This doesn't properly unlock the mutex.

wplapper replied:
code gone!

MichaelEischer commented on internal/checker/checker.go:
```diff
@@ -522,3 +536,48 @@ func (c *Checker) ReadPacks(ctx context.Context, packs map[restic.ID]int64, p *p
+					if len(result) == 0 {
+						return fmt.Errorf("checker.LookupBlob: datablob %s not mapped!", content.Str())
+					}
+					c.packSet.Insert(result[0].PackID)
```
Should insert all relevant pack files.

wplapper replied:
latest commit should fix this.

MichaelEischer commented on internal/checker/checker.go:
```diff
@@ -522,3 +536,48 @@ func (c *Checker) ReadPacks(ctx context.Context, packs map[restic.ID]int64, p *p
+	wg, wgCtx := errgroup.WithContext(ctx)
+	treeStream := restic.StreamTrees(wgCtx, wg, repo, restic.IDs{*sn.Tree}, func(tree restic.ID) bool {
+		visited := visitedTrees.Has(tree)
+		visitedTrees.Insert(tree)
```
The pack files for the tree blobs should also be checked. (see restic.FindUsedBlobs).

wplapper replied:
will check all blobs.

wplapper replied:
used restic.FindUsedBlobs

wplapper committed:
restic check with snapshot list

reworked the code using snapshotFilter.FindAll to find all snapshots and
restic.FindUsedBlobs to retrieve all used blobs.
range repo.LookupBlob (as before) to convert the blobs to their containing packfiles
and c.repo.List(ctx, restic.PackFile, ...) to retrieve the sizes of those packfiles.

Additional documentation and tests are still outstanding.

wplapper commented:
Sorry, my git history is messed up again. I don't know what I am doing wrong nor do I know of how to rectify it.

wplapper committed:
restic check with snapshots

fixed lint errors

wplapper committed:
check - rebase part 3 - add documentation and file issue-3326

add documentation for checking via snapshot filter.

cmd_check: change 1st line of issue-3326 to recognized format.

check with snapshot filter: changed issue description

wplapper commented:
A question: where do I find the `checker-test-repo.tar.gz`. I cannot find it in the `cmd/restic/testdata` directory. I would like to see how it is structured: snapshots, index, packfiles, the lot.

Thanks for your help.

MichaelEischer commented:
Fixing the branch was actually a bit of challenge this time. I had to drop all the manual merge commits of your other PRs. To sync your local state with the updated content of `check_snapshots` either clone the repository again or use the following commands. (Assumes that the git remote `origin` points to your fork of restic).
```
git fetch origin
git checkout check_snapshots
# Warning: this will delete all local changes. If necessary create a backup copy first!!!
git reset --hard origin/check_snapshots
```

The basic steps for rebasing your branch should be:
- sync master branch of fork with that of restic. The easiest way is probably to use the `Sync fork` button on https://github.com/wplapper/restic/ .
- then use the following commands:
```bash
git checkout master
git pull
# if `git pull` results in the creation of a merge commit or an error then reset your local branch as follows
# warning this deletes any local changes compared to the master branch in your fork
# `git reset --hard origin/master`
git checkout check_snapshots
# `-i` shows what will be done
git rebase -i master
```
- If a problem occurs then `git status` is usually very helpful. If the explanation there isn't sufficient for you, you could feed it to ChatGPT, which should be capable of guiding you through basic git issues.

MichaelEischer commented on internal/checker/checker.go:
```diff
@@ -522,3 +521,57 @@ func (c *Checker) ReadPacks(ctx context.Context, packs map[restic.ID]int64, p *p
+
+// CheckWithSnapshots will process snapshot IDs from command line and
+// madify c.packs so it contains only the selected packfiles via snapshotFilter
+func (c *Checker) CheckWithSnapshots(ctx context.Context, repo *repository.Repository, args []string, snapshotFilter *restic.SnapshotFilter) (bool, error) {
```
Now that I see the updated result, I think it should be refactored further into the following:

```go
func (c *Checker) PacksForSnapshots(ctx context.Context, snapshots []string, filter *restic.SnapshotFilter) (map[restic.ID]int64, error)
```

The method would no longer modify `c.packs` but only return a new map with a limited subset of packs.
That way the method forms a counterpart to `c.GetPacks()`. Then in `cmd_check.go` retrieve the packs map from either that method or `c.GetPacks()` and use that in the read data method calls.

P.S.: The `repo` parameter is redundant as the checker can use `c.repo`.

wplapper replied:
done.

MichaelEischer replied:
The changes marked as done missing.

wplapper replied:
> The changes marked as done missing.

I am currently in Germany visiting family and friends. I'll check the repository when I am home at the end of next week.

MichaelEischer reviewed (changes requested):
The changes are moving in the right direction. I've left a additional comments how to clean up the implementation further.

wplapper commented on internal/checker/checker.go:
```diff
@@ -522,3 +521,57 @@ func (c *Checker) ReadPacks(ctx context.Context, packs map[restic.ID]int64, p *p
+	}
+
+	// convert blobs to packfile IDs
+	c.packs = map[restic.ID]int64{}
```
done.

wplapper commented on internal/checker/checker.go:
```diff
@@ -522,3 +521,57 @@ func (c *Checker) ReadPacks(ctx context.Context, packs map[restic.ID]int64, p *p
+		if err != nil {
+			return err
+		} else if ctx.Err() != nil {
+			return ctx.Err()
```
else if clause removed.

wplapper commented on internal/checker/checker.go:
```diff
@@ -522,3 +521,57 @@ func (c *Checker) ReadPacks(ctx context.Context, packs map[restic.ID]int64, p *p
+		return false, err
+	}
+
+	return len(selectedTrees) > 0, nil
```
correct!

wplapper commented on internal/checker/checker.go:
```diff
@@ -522,3 +521,57 @@ func (c *Checker) ReadPacks(ctx context.Context, packs map[restic.ID]int64, p *p
+		return false, err
+	}
+
+	if len(selectedTrees) == 0 {
```
moved code back into `cmd/restic/cmd_check.go` and pass it down as a parameter to `CheckWithSnapshots`

wplapper committed:
check with snapshot filter: check early for filter errors - rebase part 4

Hand down filtered tree IDs to CheckWithSnapshots which builds 'usedBlobs'.
repo.LookupBlob is used to derive packfiles from 'usedBlobs' and constructs
'snapPacks' and overwrites c.packs.

wplapper committed:
check: cmd_check, checker, checker_test: added test - rebase part 5: add checker_test

cmd_check, checker: added error return
checker_test: testing CheckWithSnapshots

checker_test: fixed lint error for empty tree list

wplapper committed:
check: harmonize my branch with master - integrate checkSummary

based on the change in runCheck return values the following functions have been afftected
cmd/restic/cmd_check_integration_test.go
cmd/restic/cmd_migrate.go
cmd/restic/cmd_prune_integration_test.go
cmd/restic/integration_test.go

check - corrected rebase error by not removing obsolete code.

the check for the filter if len(args) > 0 { ... } is obsolete and has been replaced by
if len(args) > 0 || !opts.SnapshotFilter.Empty() { ... }

wplapper commented:
I carefully rebased this PR, following https://www.squash.io/how-to-rebase-local-branch-onto-remote-master-in-git/ . I hope my rebasing is acceptable.

wplapper committed:
restic check - add integration tests

cmd_check.go - minor changes: changes error meesage to New instead of Fatal
added a short description in the help text.

cmd_check_integration_test.go: added two tests which should both fail:
invalid input

wplapper committed:
internal/checker: added tests

checker.go: added error message if 'selectedTrees' is empty
checker_test.go: modified test checking using test.OK() and test.Assert() to
make the checking more compact.

wplapper commented:
I decided to use a github provided e-mal address and I did not see a way of changed my private e-mail address in already existing commits. Therefore I close this PR. It is continued in #5469.
